dali/util/shared_mem.cc

```diff
@@ -0,0 +1,193 @@
+  if (fd_ < 0) {
+    throw std::runtime_error("shm_open call failed");
+  }
+  ::close(temp_fd);
+  // The memory chunks will be passed between processes by sending descriptors over the sockets,
+  // not through the filenames; this way in case of unexpected process exit we're not leaking any
+  // filelike objects in /dev/shm or /tmp. The OS guarantees to keep the memory as long as
+  // any process has fds or has mmaped the memory.
+  shm_unlink(temp_filename);
+}
+
+ShmFdWrapper::~ShmFdWrapper() {
+  close();
+}
+
+int ShmFdWrapper::get_fd() {
+  return fd_;
+}
+
+int ShmFdWrapper::close() {
+  int ret = 0;
+  if (fd_ >= 0) {
+    ret = ::close(fd_);
+    fd_ = -1;
+  }
+  return ret;
+}
+
+MapMemWrapper::MapMemWrapper(int fd, uint64_t size)
+    : size_{size},
+      ptr_{static_cast<SharedMem::b_type *>(
+          mmap(nullptr, size_, PROT_WRITE | PROT_READ, MAP_SHARED, fd, 0))} {
+  if (ptr_ == MAP_FAILED) {
+    throw std::runtime_error("mmap failed");
+  }
+}
+
+MapMemWrapper::~MapMemWrapper() {
+  unmap();
+}
+
+uint8_t *MapMemWrapper::get_raw_ptr() {
+  return ptr_;
+}
+
+void MapMemWrapper::resize(uint64_t new_size) {
+  if (!ptr_) {
+    throw std::runtime_error("cannot resize the memory mapping, because no memory is mapped");
+  }
+  ptr_ = static_cast<SharedMem::b_type *>(mremap(ptr_, size_, new_size, MREMAP_MAYMOVE));
+  if (ptr_ == MAP_FAILED) {
+    throw std::runtime_error("mremap failed");
+  }
+  size_ = new_size;
+}
+
+int MapMemWrapper::unmap() {
+  int ret = 0;
+  if (ptr_) {
+    ret = munmap(ptr_, size_);
+    ptr_ = nullptr;
+  }
+  return ret;
+}
+
+SharedMem::SharedMem(int fd, uint64_t size) : size_{size * sizeof(SharedMem::b_type)} {
+  if (fd >= 0) {
+    fd_ = std::make_unique<ShmFdWrapper>(fd);
+  } else {
+    fd_ = std::make_unique<ShmFdWrapper>();
+    if (ftruncate(fd_->get_fd(), size_) == -1) {
+      throw std::runtime_error("failed to resize shared memory");
+    }
+  }
+  mem_ = std::make_unique<MapMemWrapper>(fd_->get_fd(), size_);
+}
+
+uint64_t SharedMem::size() const {
+  return size_;
+}
+
+int SharedMem::fd() {
+  return !fd_ ? -1 : fd_->get_fd();
+}
+
+SharedMem::b_type *SharedMem::get_raw_ptr() {
+  return !mem_ ? nullptr : mem_->get_raw_ptr();
+}
+
+void SharedMem::resize(uint64_t size, bool trunc) {
+  size_ = size * sizeof(SharedMem::b_type);
+  if (trunc) {
+    if (ftruncate(fd_->get_fd(), size_) == -1) {
+      throw std::runtime_error("failed to resize shared memory");
+    }
+  }
+  if (mem_) {
+    mem_->resize(size_);
+  } else {
+    if (!fd_) {
+      throw std::runtime_error("cannot mmap memory - no file descriptor");
+    }
+    mem_ = std::make_unique<MapMemWrapper>(fd_->get_fd(), size_);
+  }
+}
+
+  void SharedMem::close_fd() {
+    if (fd_ && fd_->close() != 0) {
+      throw std::runtime_error("closing fd failed");
+    }
+  }
+
+  void SharedMem::close_map() {
+    if (mem_ && mem_->unmap() != 0) {
+      throw std::runtime_error("unmaping shared memory failed");
```

Comment:
Some error description or at least error code would be nice... or more than nice.

Reply:
done